src/data/matrix/basic.lean

```diff
@@ -26,24 +26,120 @@ open dmatrix
 
+/-- `equiv.perm.to_matrix` converts a permutation `σ : equiv.perm m` to a matrix
+    containing ones and zeros. This definition is then used to define the Prop `matrix.is_perm`.
+    `equiv.perm.to_prequiv_to_matrix_eq_to_matrix` proves that
+    `σ.to_pequiv.to_matrix = σ.to_matrix` in `pequiv.lean`.-/
+def equiv.perm.to_matrix [decidable_eq m] [has_zero α] [has_one α] (σ : equiv.perm m) :
+matrix m m α
+| i j := if σ i = j then 1 else 0
+
 namespace matrix
 
+instance [inhabited α] : inhabited (matrix m n α) := pi.inhabited _
+instance [has_add α] : has_add (matrix m n α) := pi.has_add
+instance [add_semigroup α] : add_semigroup (matrix m n α) := pi.add_semigroup
+instance [add_comm_semigroup α] : add_comm_semigroup (matrix m n α) := pi.add_comm_semigroup
+instance [has_zero α] : has_zero (matrix m n α) := pi.has_zero
+instance [add_monoid α] : add_monoid (matrix m n α) := pi.add_monoid
+instance [add_comm_monoid α] : add_comm_monoid (matrix m n α) := pi.add_comm_monoid
+instance [has_neg α] : has_neg (matrix m n α) := pi.has_neg
+instance [has_sub α] : has_sub (matrix m n α) := pi.has_sub
+instance [add_group α] : add_group (matrix m n α) := pi.add_group
+instance [add_comm_group α] : add_comm_group (matrix m n α) := pi.add_comm_group
+instance [unique α] : unique (matrix m n α) := pi.unique
+instance [subsingleton α] : subsingleton (matrix m n α) := pi.subsingleton
+instance [nonempty m] [nonempty n] [nontrivial α] : nontrivial (matrix m n α) :=
+function.nontrivial
+
+instance [has_scalar R α] : has_scalar R (matrix m n α) := pi.has_scalar
+instance [has_scalar R α] [has_scalar S α] [smul_comm_class R S α] :
+  smul_comm_class R S (matrix m n α) := pi.smul_comm_class
+instance [has_scalar R S] [has_scalar R α] [has_scalar S α] [is_scalar_tower R S α] :
+  is_scalar_tower R S (matrix m n α) := pi.is_scalar_tower
+instance [monoid R] [mul_action R α] :
+  mul_action R (matrix m n α) := pi.mul_action _
+instance [monoid R] [add_monoid α] [distrib_mul_action R α] :
+  distrib_mul_action R (matrix m n α) := pi.distrib_mul_action _
+instance [semiring R] [add_comm_monoid α] [module R α] :
+  module R (matrix m n α) := pi.module _ _ _
+
 section ext
+
 variables {M N : matrix m n α}
 
 theorem ext_iff : (∀ i j, M i j = N i j) ↔ M = N :=
 ⟨λ h, funext $ λ i, funext $ h i, λ h, by simp [h]⟩
 
-@[ext] theorem ext : (∀ i j, M i j = N i j) → M = N :=
-ext_iff.mp
+@[ext] theorem ext : (∀ i j, M i j = N i j) → M = N := ext_iff.mp
 
 end ext
 
+section transpose
+
+/-- The transpose of a matrix. -/
+def transpose (M : matrix m n α) : matrix n m α
+| x y := M y x
+localized "postfix `ᵀ`:1500 := matrix.transpose" in matrix
+
+lemma trans_col_eq_row (A : matrix m n α) (i : m) : (λ j, Aᵀ j i) = A i :=
+by simp [transpose]
+lemma trans_row_eq_col (A : matrix m n α) (j : n) : Aᵀ j = (λ i, A i j):=
+by ext; simp [transpose]
+
+instance vec_has_star [has_star α]: has_star (n → α) := ⟨λ v i, star (v i)⟩
```

Comment:
I'm pretty sure this instance already exists

Reply:
`infer_instance` fails for me.

Reply:
It's here: https://leanprover-community.github.io/mathlib_docs/algebra/star/pi.html#pi.has_star

Reply:
To use this. I think I need to import `import algebra.star.pi`. But it seems that this cannot be imported into the current file.

Reply:
That's not an excuse to avoid using the instance I'm afraid - instead that tells you that either you need to put this stuff in another file, or we need to remove some imports from `pi.has_star`.

Reply:
#8303 fixes this

Reply:
Should I wait for #8303 to merge first? @eric-wieser

Reply:
There's not really any reason too - the rest of this PR doesn't use this instance, so you could just as easily just remove it from your PR.

Reply:
This blocks `conj_transpose_col_eq_star_row` and same for the row. But perhaps not add these two in this PR.

Reply:
I can add them later if they are worth adding.

Reply:
I added those two lemmas to #8303, thanks for reminding me of them.

Reply:
Yeah, I added the transpose ones in this PR.